introduce a "tree" type to the flattenSpec

jasonk000 opened the pull request:
### Overview and benefits

This introduces a `tree` to the `flattenSpec` which allows faster JSON parsing than `jq` and `path` parsing types in the event that a simple hierarchical lookup is required.

For one of our high-volume ingestion tasks, this reduces middle manager CPU load by 25%. This task presently has 3 `root` fields and 8 additional `path` fields of the form 

```
{
  "type": "path",
  "name": "task_detail",
  "expr": "$.metadata.detail"
},
```

Replacing this with a hierarchical lookup as follows:
```
{
  "type": "tree",
  "name": "task_detail",
  "exprs": ["metadata", "detail"]
},
```

In this sample, a switch to the new `tree` spec is submitted at 15.54, and after a moment or two the CPU shifts down at a notably lower total utilization from 3 cores to 2.3 cores.
![image](https://user-images.githubusercontent.com/3196528/150242320-61fedc3d-3c1f-419d-a6b6-273cc47658a8.png)


### Implementation decision

The goal is to introduce the ability to (1) provide a list of expressions instead of a single expression, and, a new `tree` type. 
```
{
  "type": "tree",                     <-- new type
  "name": "task_detail",
  "exprs": ["metadata", "detail"]     <-- name of json keys to traverse as a List<String>
},
```

The crux of the change is this tree extractor for loop (below) and the requisite parser configuration to ensure it is called.
```
public Function<JsonNode, Object> makeJsonTreeExtractor(final List<String> exprs)
{
  String[] exprsArray = exprs.toArray(new String[0]);
  return node -> {
    JsonNode target = node;
    for (String expr : exprsArray) {
      target = target.get(expr);
      if (target == null) {
        break;
      }
    }
    return valueConversionFunction(target);
  };
}
```

Key consideration was to avoid any backwards compatibility issues with deployments, so I opted for using a separate `exprs` parameter rather than overload the `expr` parameter.

### Changes required

The change adds enum entries, the required field spec to parse `List<String> exprs` to configuration, and validate that `exprs` is only provided for the `tree` type.

Then, tie it all into the `ObjectFlatteners` and `JSONFlattenerMaker` with a new `makeJsonTreeExtractor` that iterates each of the keys as a hierarchy in the dataset.

#### Supporting changes

- Update a lot of tests to explicitly verify the parameters are loaded.
- Update `docs/ingestion-data-formats.md` to include a description of the `tree` type and the `exprs` parameter.
- Update Avro, Orc, and Parquet flatteners to acknowledge `tree` but only work with a depth of 1; (would welcome guidance here).
- Added benchmarks to `FlattenJSONBenchmark` to confirm improvements in flattener layer

### Additional supporting data

On flamegraphs, `Flattener` related code before the change makes up ~49% of the `task-runner-xx` threads utilisation; after the change only ~16% of the `task-runner-xx` thread is used running `Flattener` related code.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] N/A added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] N/A added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

jasonk000 commented:
I haven't added an integration test as I think the existing tests cover it sufficiently, but I am OK to do so if this change warrants it.

FrankChen021 commented on core/src/main/java/org/apache/druid/java/util/common/parsers/JSONPathFieldSpec.java:
```diff
@@ -23,32 +23,55 @@
+      case TREE:
+        this.expr = null;
+        this.exprs = Preconditions.checkNotNull(exprs, "Missing 'exprs' for field[%s]", name);
+        Preconditions.checkArgument(this.exprs.size() >= 1, "Need at least one 'exprs' value for field[%s]", name);
```
can use `CollectionsUtils.isNullOrEmpty` to combine these two checks together.

jasonk000 replied:
thanks for the tip, fixed in 672e039e519728bde493bf2ab076df9d586db4f9.

FrankChen021 commented on core/src/main/java/org/apache/druid/java/util/common/parsers/JSONPathFieldSpec.java:
```diff
@@ -23,32 +23,55 @@
       @JsonProperty("name") String name,
-      @JsonProperty("expr") String expr
+      @JsonProperty("expr") String expr,
+      @JsonProperty("exprs") List<String> exprs
```
`exprs` and `expr` may be ambiguous. I think maybe we can name it as `nodes` because node is a concept related to a tree.

jasonk000 replied:
fixed in 672e039e519728bde493bf2ab076df9d586db4f9.

FrankChen021 commented on core/src/main/java/org/apache/druid/java/util/common/parsers/JSONFlattenerMaker.java:
```diff
@@ -108,6 +108,22 @@ public Object getRootField(final JsonNode obj, final String key)
+  @Override
+  public Function<JsonNode, Object> makeJsonTreeExtractor(final List<String> exprs)
+  {
+    String[] exprsArray = exprs.toArray(new String[0]);
```
Why do we need to get the string array first? Can't we use the `exprs` object directly in the loop below?

jasonk000 replied:
This is a defensive copy.

FrankChen021 commented on docs/ingestion/data-formats.md:
```diff
@@ -606,11 +606,12 @@ Flattening is only supported for [data formats](data-formats.md) that support ne
+| type  | Options are as follows:<br><br><ul><li>`root`, referring to a field at the root level of the record. Only really useful if `useFieldDiscovery` is false.</li><li>`path`, referring to a field using [JsonPath](https://github.com/jayway/JsonPath) notation. Supported by most data formats that offer nesting, including `avro`, `json`, `orc`, and `parquet`.</li><li>`jq`, referring to a field us ...
+| name  | Name of the field after flattening. This name can be referred to by the [`timestampSpec`](./ingestion-spec.md#timestampspec), [`transformSpec`](./ingestion-spec.md#transformspec), [`dimensionsSpec`](./ingestion-spec.md#dimensionsspec), and [`metricsSpec`](./ingestion-spec.md#metricsspec).                                                                                                     ...
+| expr  | Expression for accessing the field while flattening. For type `path`, this should be [JsonPath](https://github.com/jayway/JsonPath). For type `jq`, this should be [jackson-jq](https://github.com/eiiches/jackson-jq) notation. For other types, this parameter is ignored.                                                                                                                          ...
+| exprs | For `tree` only. Multiple-expression field for accessing the field while flattening, representing the hierarchy of field names to read. For other types, this parameter must not be provided.                                                                                                                                                                                                         ...
```
Since `avro`, `parquet`, `orc` format is not fully supported by this new feature, it's better to document these constraints here.

jasonk000 replied:
Fixed in 429fd0d13a.

I've added some extras specifically in the Avro, ORC, and Parquet sections.

Also I'm not familiar with Avro, ORC, or Parquet, but would welcome any guidance on how to correctly implement this as I think the implementation could be simple enough.

jasonk000 replied:
Also fixed a typo in 4f80239.

FrankChen021 commented on core/src/main/java/org/apache/druid/java/util/common/parsers/JSONFlattenerMaker.java:
```diff
@@ -108,6 +108,22 @@ public Object getRootField(final JsonNode obj, final String key)
+      for (String keyName : keyNames) {
+        targetNode = targetNode.get(keyName);
+        if (targetNode == null) {
+          break;
```
We can directly return null here.

jasonk000 replied:
fixed in 9b283122f9

jasonk000 commented on core/src/main/java/org/apache/druid/java/util/common/parsers/JSONFlattenerMaker.java:
```diff
@@ -108,6 +108,22 @@ public Object getRootField(final JsonNode obj, final String key)
+  @Override
+  public Function<JsonNode, Object> makeJsonTreeExtractor(final List<String> nodes)
+  {
+    String[] keyNames = nodes.toArray(new String[0]);
```
It's a defensive copy in case the caller does something to change the `List<>`. It does not need to be to an array if a List is preferred.

jasonk000 replied:
added a comment to explain in 9b283122f9

jasonk000 commented:
Hi @FrankChen021 , looks like you've approved. What's next to merge it?

FrankChen021 commented:
@jasonk000 Could you take a look at the CI?

abhishekagarwal87 commented:
@jasonk000 - can you look at these build failures?

jasonk000 commented:
Yes, @FrankChen021 & @abhishekagarwal87 this is fixed now.

(Squashed & rebased, and incorporate changes from e5ad862665d.)

abhishekagarwal87 commented on core/src/main/java/org/apache/druid/java/util/common/parsers/ObjectFlatteners.java:
```diff
@@ -231,6 +235,11 @@ public Map<String, Object> toMap(T obj)
+    /**
+     * Create a "field" extractor for nested json expressions
+     */
+    Function<T, Object> makeJsonTreeExtractor(List<String> nodes);
```
@jasonk000 @FrankChen021 - this can potentially break extensions developed outside apache repo. While the interface is not marked `ExtensionPoint`, it does get extended outside the core. My suggestion will be to throw an `UnsupportOperationException` in a default implementation. Also `getRootField` needs to be called as before when `fieldSpec.getType()` returns `ROOT`. 
what do you think?

FrankChen021 replied:
Do we need to care about an interface which is not marked as `ExtensionPoint`? Although we can add a default implementation for this interface. Another question is should we mark it as `ExtensionPoint`?

abhishekagarwal87 replied:
It's already implemented in Avro, orc, and parquet extension. So +1 to both of your suggestions. We should add a default impl and and mark it extension point.

FrankChen021 replied:
Hi @jasonk000 , could you make some changes here according to above opinions? I think we're close to get this PR to be merge.

jasonk000 replied:
OK let me take a look. I'm a little hesitant as I'm not in a position to _run_ it now against trunk+this rebased, but I'll put up the changes and we can take it from there!

abhishekagarwal87 commented on docs/ingestion/data-formats.md:
```diff
@@ -254,11 +254,11 @@ To use the Parquet input format load the Druid Parquet extension ([`druid-parque
+| Field | Type | Description                                                                                                                                                   | Required |
+|-------|------|---------------------------------------------------------------------------------------------------------------------------------------------------------------|----------|
+|type| String| Set value to `parquet`.                                                                                                                                       | yes |
+|flattenSpec| JSON Object | Define a [`flattenSpec`](#flattenspec) to extract nested values from a Parquet file. Only 'path' expressions are supported ('jq' and 'tree' are unavailable). | no (default will auto-discover 'root' level properties) |
```
can you a similar comment for ORC too?

abhishekagarwal87 reviewed (commented):
It is mostly LGTM except for potential incompatibility. I am also wondering if `tree` is the right name for this new technique. To me, it looks very similar to `path` except we are passing a list of nodes instead of passing a dot-separated string. what if we do not introduce a separate kind of field type. Instead, we chose this new more efficient implementation for JsonFlattenMaker if the path is a simple form of `a.b.c`?  The optimization can be turned on and off by the user.

jasonk000 commented:
I've rebased and addressed the changes requested, pending a build.

FWIW, this was originally built for v20, and I haven't had an opportunity to run this full change through production over a v25 so I cannot confidently say what the performance improvement currently is.

FrankChen021 commented:
Thank you @jasonk000 I will review it in this week. 

BTW, it's better to merge the master back to your own branch instead of rebasing and force push because the former one would ease some review work on GitHub for reviewers.

FrankChen021 reviewed (approved):
LGTM. Thank you @jasonk000 
@abhishekagarwal87 do you have any other comments?

abhishekagarwal87 reviewed (approved):
LGTM too. Thank you, @jasonk000

jasonk000 commented:
Great! What's needed to merge it?

FrankChen021 commented:
I merged it. Thanks for the contribution @jasonk000
